### DISCUSSION TITLE
How do I handle views in associated schemas?

### DISCUSSION THREAD

**Discussion by arogozhnikov** [sqlalchemy/sqlalchemy#9770] [Usage Questions]
Thanks for sqlalchemy/alembic! I deal with a specific usage scenario, maybe you have some ideas how to implement that.

I have two postgresql schemas - live and dev, corresponding to live stage and dev stage.
In addition I have two schemas for external collaborators: live-ext and dev-ext.
Every table in live has a corresponding view in live-ext, which restricts original table to share only specific records. Same for dev and dev-ext. External collaborators see data in real time, but filtered to only allowed records, and have no way to edit the data.

When organized this way, sqlalchemy can query from any schema: live or live-ext using the same definition of ORM classes. This is very convenient.

But migration now became hard.

My previous pipeline was this: I could auto-generate revisions with alembic on dev, then apply it in live (alembic just logs under a different user to get access to a different stage, commands are not qualified, so migration commands can be safely applied).

I can use alembic_utils to manage views, but when autogeneration happens, it needs to fully qualify view, in this case view will be generated on 'dev-ext' schema, and migration can't be applied to live & live-ext.

My current (inefficient) pipeline is: before migration I tear down dev-ext manually, apply alembic migration to dev, create dev-ext manually. Same for live. 

I guess this can be done better? 😄

**Comment by zzzeek** [comment]
I dont think I'm following this very specifically, but usually migrations that affect multiple schemas would be written that way, if you want to migrate tables and views in schemas A and B, the alembic `op.create_xyz()` commands would individually refer to "A" and "B" for their schemas.    I havent used alembic_utils but IIUC it just generates autogen commands for views, which is helpful.    There wouldn't be any manual steps once the migrations are written.   

guess im not really following the problem or what you mean by "create dev-ext manually"

**Reply by arogozhnikov** [reply]
so, the issue is if migration has something like
```sql
create view dev-ext.users as ( select * from dev.users where <user in same group or some other condition> )
```

then the same migration in live looks like
```sql
create view live-ext.users as ( select * from live.users where <user in same group or some other condition> )
```

and having the same migration working in these two different ways is hard.


> what you mean by "create dev-ext manually

I just have a cli command that drops all views in dev-ext, and another cli command that creates all of them, this command is just a set of sql commands, not tied to alembic

**Reply by zzzeek** [reply]
We have a feature called ["schema translate"](https://docs.sqlalchemy.org/en/20/core/connections.html#schema-translating) for this.   Alembic is not too well integrated with it for fine grained operations.  however, I'd use an approach like that.   code the migrations using symbolic names and then change the names "dev-ext" and "live-ext" on the fly as the statements are executed.  it depends on how you are running those "create view" statements.
